app/components/Taps/TapConfiguration/S3/index.js

```diff
@@ -0,0 +1,350 @@
+/*
+ * knots
+ * Copyright 2018 data.world, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the
+ * License.
+ *
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * This product includes software developed at
+ * data.world, Inc.(http://data.world/).
+ */
+
+// @flow
+/* eslint-disable react/prop-types */
+/* eslint-disable camelcase */
+
+import React, { Component } from 'react';
+import {
+  Col,
+  Container,
+  Form,
+  FormFeedback,
+  FormGroup,
+  FormText,
+  Input,
+  Label,
+  Row,
+  Button,
+  Alert,
+  Table
+} from 'reactstrap';
+import { WithContext as ReactTags } from 'react-tag-input';
+import styles from './S3.css';
+
+const KeyCodes = {
+  comma: 188,
+  enter: 13,
+  tab: 9
+};
+const delimiters = [KeyCodes.comma, KeyCodes.enter, KeyCodes.tab];
+
+type Props = {
+  tapsStore: {
+    'tap-s3-csv': {
+      fieldValues: {
+        bucket: string,
+        start_date: string,
+        tables: string
+      }
+    }
+  },
+  updateTapField: (tap: string, field: string, value: string | number) => void
+};
+type State = {
+  bucket: '',
+  start_date: '',
+  tables: [{ table_name: '', key_properties: [], search_pattern: '' }],
+  tags: []
+};
+
+export default class S3 extends Component<Props, State> {
+  constructor(props) {
+    super(props);
+    this.state = {
+      bucket: '',
+      start_date: '',
+      tables: [{ table_name: '', key_properties: [], search_pattern: '' }],
+      tags: []
+    };
+  }
+
+  componentWillMount = () => {
+    const tagObj = {};
+    const tagsArr = [];
+    const { tables } = this.props.tapsStore['tap-s3-csv'].fieldValues;
+    const tableValue = tables
+      ? JSON.parse(tables)
+      : [{ table_name: '', key_properties: [], search_pattern: '' }];
+    tableValue.forEach((key) => {
+      const keyProperties = key.key_properties;
+      if (keyProperties.length !== 0) {
+        tagObj.id = keyProperties;
+        tagObj.text = keyProperties;
+        tagsArr.push(tagObj);
+      }
+    });
+    this.setState({
+      tables: tableValue,
+      tags: tagsArr
+    });
+  };
+
+  validate = (field: string, value: string) => {
+    if (value) {
+      this.setState({ [field]: { valid: true } });
+    } else {
+      this.setState({ [field]: { invalid: true } });
+    }
+  };
+
+  toISODateString = (date: Date) => {
```

Comment:
This has a bug that @kndungu fixed for other taps. Sad to see it come back. Another DRY opportunity.